url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/ops/MatrixIO.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (c) 2009-2013, Peter Abeles. All Rights Reserved.
+ *
+ * This file is part of Efficient Java Matrix Library (EJML).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ejml.ops;
+
+import org.ejml.data.DenseMatrix64F;
+import org.ejml.data.Matrix64F;
+import org.ejml.data.ReshapeMatrix64F;
+
+import java.io.*;
+
+
+/**
+ * Provides simple to use routines for reading and writing matrices to and from files.
+ *
+ * @author Peter Abeles
+ */
+public class MatrixIO {
+
+    /**
+     * Saves a matrix to disk using Java binary serialization.
+     *
+     * @param A The matrix being saved.
+     * @param fileName Name of the file its being saved at.
+     * @throws java.io.IOException
+     */
+    public static void saveBin(ReshapeMatrix64F A, String fileName)
+        throws IOException
+    {
+        FileOutputStream fileStream = new FileOutputStream(fileName);
+        ObjectOutputStream stream = new ObjectOutputStream(fileStream);
+
+        try {
+            stream.writeObject(A);
+            stream.flush();
+        } finally {
+            // clean up
+            try {
+                stream.close();
+            } finally {
+                fileStream.close();
+            }
+        }
+
+    }
+
+    /**
+     * Loads a DeneMatrix64F which has been saved to file using Java binary
+     * serialization.
+     *
+     * @param fileName The file being loaded.
+     * @return  DenseMatrix64F
+     * @throws IOException
+     */
+    public static <T extends ReshapeMatrix64F> T loadBin(String fileName)
+        throws IOException
+    {
+        FileInputStream fileStream = new FileInputStream(fileName);
+        ObjectInputStream stream = new ObjectInputStream(fileStream);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *stream*. The resource is referenced at lines 74, 78, 79. The resource is closed at line 86. There are other execution paths that do not contain closure statements, for example, when an if condition at line 79 evaluates to true. Ensure *stream* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).